README.md
# SurrealDB Test
## Setting Up SurrealDB
### Follow the installation guide at https://surrealdb.com/docs/installation for your OS
After installing SurrealDB, start the server with:</br>
`surreal start --log trace --user root --pass root memory`

## Setting Up Backend
Open a new terminal</br>
Navigate to the root directory of the backend project</br>
Install backend dependencies</br>
`npm install`

## Running Backend
Make sure you are in the root directory of the backend project<br>
Start the backend server<br>
`npm run start`

## Setting Up Frontend
Open a new terminal</br>
Navigate to the frontend directory inside the project</br>
`cd frontend`
Install frontend dependencies<br>
`npm install`

## Running Frontend
Navigate to the frontend directory<br>
`cd frontend`
Start the frontend server<br>
`npm run start`

## Now, both backend and frontend servers should be running. 
You can use the web application in your browser to interact with the database through the backend API.

## Usage
With both the backend and frontend servers running, you can use the web application in your browser to interact with the database through the backend API. The application should allow you to register, login, create posts, like posts, and comment on posts.


history.txt
#V2
select * from user;
select * from post;
select * from user:user1
select * from user:user1->like->post:l5338fjlzgoswyet9axh
select * from like
select * from like where in = user:user1
remove db like
remvoe table like
select * from like
remove table like
select * from like
remove table company
INSERT INTO company (name, founded) VALUES ('SurrealDB', '2021-09-10');
define field value on company flexible type object;
INSERT INTO company (name, founded, value) VALUES ('MongoDB', '2021-09-10', 6);
INSERT INTO company (value) VALUES (6);
DEFINE company schemafull
DEFINE company schemaful
DEFINE table company schemaful
define field value on company flexible type object;
INSERT INTO company (value) VALUES (6);
select * from company
remove table company
INSERT INTO company (name, founded) VALUES ('SurrealDB', '2021-09-10');
select * from company;
insert into company (testField) values (1) 
select * from company;
insert into company (testField2) values ('test') 
define field testfield3 on company type string;
insert into company (testField3) values ('test') 
insert into company (testField2) values ('test') 
insert into company (testField3) values (5) 
remove table company
select * from company;
insert into company (name) values ('test')
insert into company (age) values (35)
define table comapny schemafull
define field testfield1 type string;
define field testfield1 on company type string;
insert into company (name) values ('test1')
select * from company;
define field name on company type number;
insert into company (name) values ('test2')
insert into company (name) values (10)
define field age on company type string;
insert into company (age) values ('ten')
insert into company (age) values (10)
insert into company (age) values ({age: 10})
insert into company (name) values ({age: 10})
insert into company (field1) values ('value')
select * from company
insert into company (field1) values ({age: 10})
define field1 age on company type number;
define field field1 on company type number;
insert into company (field1) values ({age: 10})
define table company schemafull
insert into company (field2) values (10)
insert into company (test) values (10)
define field field2 on company type number;
insert into company (field2) values (10)
insert into company (field2, field3) values (10, 13)
remove table company


frontend/src/components/Post.js
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { useParams, useNavigate } from 'react-router-dom';

const instance = axios.create({
  baseURL: 'http://localhost:4000',
});

function Post() {
  const [post, setPost] = useState(null);
  const [comments, setComments] = useState([]);
  const [newComment, setNewComment] = useState('');

  // Additional state for post update
  const [isEditing, setIsEditing] = useState(false);
  const [updatedPostContent, setUpdatedPostContent] = useState('');

  const { postId } = useParams();
  const username = localStorage.getItem('user');
  const navigate = useNavigate();


  const getPost = (postId) => {
    instance.get(`/post/${postId}`).then(res => {
      console.log(res);
      if (res.status === 200)
        setPost(res.data[0])
    }).catch(err => console.log(err));
  }

  useEffect(() => {
    getPost(postId);
  }, [postId]);

  const submitComment = () => {
    instance.post('/comment', { postId: postId, username, content: newComment }).then(() => {
      setNewComment('');
      instance.get(`/post/${postId}/comments`).then(res => setComments(res.data));
    });
  };

  const likePost = () => {
    instance.post('/like', { username, postId }).then(res => {
      if (res.status === 200) {
        alert('You liked this post!');
      } else {
        alert('You have already liked this post');
      }
    }).catch(err => {
      alert(err.response.data.message);
    });
  };

  // Update post function
  const updatePost = () => {
    instance.put(`/post/${postId}`, { content: updatedPostContent }).then(res => {
      if (res.status === 200) {
        alert('Post updated successfully');
        getPost(postId);
        setIsEditing(false);
        setPost(res.data[0]);
      }
    }).catch(err => alert(err.response.data.message));
  };

  // Delete post function
  const deletePost = () => {
    if (window.confirm('Are you sure you want to delete this post?')) {
      instance.delete(`/post/${postId}`).then(res => {
        if (res.status === 200) {
          alert('Post deleted successfully');
          navigate('/posts') // Redirect to home after deletion
        }
      }).catch(err => alert(err.response.data.message));
    }
  };

  if (!post) return 'Loading...';

  return (
    <div className="p-8">
      {
        isEditing ? (
          <>
            <textarea
              className="w-full p-2 mb-4 border rounded-md"
              value={updatedPostContent}
              onChange={e => setUpdatedPostContent(e.target.value)}
            />
            <button
              className="mb-4 mr-2 p-2 font-semibold text-white bg-blue-600 rounded-md hover:bg-blue-700"
              onClick={updatePost}
            >
              Save Changes
            </button>
            <button
              className="mb-4 p-2 font-semibold text-white bg-red-600 rounded-md hover:bg-red-700"
              onClick={() => setIsEditing(false)}
            >
              Cancel
            </button>
          </>
        ) : (<>

          <h1 className="mb-4 text-2xl font-semibold text-gray-800">{post.content}</h1>
          <button
            className="mb-4 mr-2 p-2 font-semibold text-white bg-blue-600 rounded-md hover:bg-blue-700"
            onClick={likePost}
          >
            Like
          </button>
          <button
            className="mb-4 mr-2 p-2 font-semibold text-white bg-green-600 rounded-md hover:bg-green-700"
            onClick={() => setIsEditing(true)}
          >
            Edit
          </button>
          <button
            className="mb-4 p-2 font-semibold text-white bg-red-600 rounded-md hover:bg-red-700"
            onClick={deletePost}
          >
            Delete
          </button>
          {comments.map(comment => (
            <div key={comment.id} className="p-4 mb-2 bg-white rounded shadow">
              {comment.content}
            </div>
          ))}
          <input
            className="w-full p-2 mt-2 border rounded-md"
            value={newComment}
            onChange={e => setNewComment(e.target.value)}
            placeholder="New comment"
          />
          <button
            className="w-full p-2 mt-2 font-semibold text-white bg-blue-600 rounded-md hover:bg-blue-700"
            onClick={submitComment}
          >
            Submit comment
          </button>
        </>)
      }
    </div>
  );
}

export default Post;


frontend/src/components/Login.js
import React, { useState } from 'react';
import axios from 'axios';
import { useNavigate } from 'react-router-dom';

const instance = axios.create({
  baseURL: 'http://localhost:4000',
});

function Login() {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const navigate = useNavigate();

  const login = async () => {
    try {
      const result = await instance.post('/login', { username, password });
      if (result.status === 200) {
        navigate('/posts');
        localStorage.setItem('user', result.data.data.id);
      }
      else {
        alert('Error logging in');
      }
    }
    catch (err) {
      if (err.response.status === 401) {
        alert(err.response.data.message);
      }
      else {
        alert('Error logging in');
      }
    }
  };

  return (
    <div className="flex flex-col items-center justify-center min-h-screen bg-gray-100">
      <div className="p-4 bg-white rounded shadow-xl">
        <h2 className="text-2xl font-semibold text-center text-gray-800">Login</h2>
        <input 
          className="w-full p-2 mt-2 border rounded-md"
          value={username}
          onChange={e => setUsername(e.target.value)}
          placeholder="Username"
        />
        <input 
          className="w-full p-2 mt-2 border rounded-md"
          value={password}
          onChange={e => setPassword(e.target.value)}
          placeholder="Password"
          type="password"
        />
        <button 
          className="w-full p-2 mt-2 font-semibold text-white bg-blue-600 rounded-md hover:bg-blue-700"
          onClick={login}
        >
          Login
        </button>
      </div>
    </div>
  );
}

export default Login;


frontend/src/components/Posts.js
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { Link, useNavigate } from 'react-router-dom';

const instance = axios.create({
  baseURL: 'http://localhost:4000',
});

function Posts() {
  const [posts, setPosts] = useState([]);

  const navigate = useNavigate();

  useEffect(() => {
    instance.get('/posts').then(res => {
      if (res.status === 200) {
        setPosts(res.data[0].result);
      }
    });
  }, []);

  return (
    <div className="p-8">
      <h2 className="mb-4 text-2xl font-semibold text-gray-800">Posts</h2>
      <div className="space-y-4">
        {posts.map(post => (
          <div key={post.id} className="p-4 bg-white rounded shadow">
            <Link to={`/post/${post.id}`} className="text-blue-600 hover:underline">
              {post.content}
            </Link>
          </div>
          
        ))}
      </div>
      <button
            className="w-full p-2 mt-2 font-semibold text-white bg-blue-600 rounded-md hover:bg-blue-700"
            onClick={() => navigate('/create-post')}
          >
            Create Post
          </button>
    </div>
  );
}

export default Posts;


frontend/src/components/Register.js
import React, { useState } from 'react';
import axios from 'axios';
import { useNavigate } from 'react-router-dom';

const instance = axios.create({
  baseURL: 'http://localhost:4000',
});

function Register() {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const navigate = useNavigate();

  const register = async () => {
    try {
      const result = await instance.post('/register', { username, password });
      if (result.status === 200) {
        navigate('/login');
      } else {
        alert('Error registering');
      }
    }
    catch (err) {
      if (err.response.status === 400) {
        alert(err.response.data.message);
      }
      else alert('Error registering');
    }
  };

  return (
    <div className="flex flex-col items-center justify-center min-h-screen bg-gray-100">
      <div className="p-4 bg-white rounded shadow-xl">
        <h2 className="text-2xl font-semibold text-center text-gray-800">Register</h2>
        <input 
          className="w-full p-2 mt-2 border rounded-md"
          value={username}
          onChange={e => setUsername(e.target.value)}
          placeholder="Username"
        />
        <input 
          className="w-full p-2 mt-2 border rounded-md"
          value={password}
          onChange={e => setPassword(e.target.value)}
          placeholder="Password"
          type="password"
        />
        <button 
          className="w-full p-2 mt-2 font-semibold text-white bg-blue-600 rounded-md hover:bg-blue-700"
          onClick={register}
        >
          Register
        </button>
      </div>
    </div>
  );
}

export default Register;


frontend/src/components/CreatePost.js
import React, { useState } from 'react';
import axios from 'axios';
import { useNavigate } from 'react-router-dom';

const instance = axios.create({
  baseURL: 'http://localhost:4000',
});

function CreatePost() {
  const [content, setContent] = useState('');
  // get logged user
  const username = localStorage.getItem('user');
  const navigate = useNavigate();

  const createPost = async () => {
    try {
      await instance.post('/post', { username, content });
      navigate('/posts');
    }
    catch (err) {
      console.log(err);
    }
  };

  return (
    <div className="p-8">
      <textarea 
        className="w-full p-2 mb-4 border rounded-md"
        value={content} 
        onChange={e => setContent(e.target.value)} 
        placeholder="New post" 
      />
      <button 
        className="w-full p-2 font-semibold text-white bg-blue-600 rounded-md hover:bg-blue-700"
        onClick={createPost}
      >
        Create post
      </button>
    </div>
  );
}

export default CreatePost;


app.js
const express = require('express');
const { default: Surreal } = require('surrealdb.js');
const bodyParser = require('body-parser');
const cors = require('cors');

// Connecting to SurrealDB
const db = new Surreal('http://127.0.0.1:8000/rpc');
db.signin({
    user: 'root',
    pass: 'root',
});
db.use('app', 'app');

const app = express();
app.use(cors());
app.use(bodyParser.json());

app.get('/', (req, res) => res.send('Hello World!'));

// Route for user registration
app.post('/register', async (req, res) => {
    const { username, password } = req.body;
    const userAlready = await db.select(`user:${username}`);
    if (userAlready) {
        res.status(400).json({ message: 'User already exists', data: null });
        return;
    }
    else {
        const user = await db.create('user', { id: `user:${username}`, password });
        res.json(user);
    }
});

// Route for user login
app.post('/login', async (req, res) => {
    const { username, password } = req.body;
    const user = await db.select(`user:${username}`);
    console.log(user);
    if (user && user.password === password) {
        // This is a simplified example, normally you should use JWT or similar for handling user sessions
        res.json({ message: 'Logged in successfully', data: user });
    } else {
        res.status(401).json({ message: 'Invalid username or password', data: null });
    }
});

// Route for getting all posts
app.get('/posts', async (req, res) => {
    const posts = await db.query('SELECT * FROM post');
    res.json(posts);
});

// Route for getting a single post
app.get('/post/:postId', async (req, res) => {
    const { postId } = req.params;
    // console.log(postId);
    const post = await db.query(`SELECT * FROM post WHERE id = ${postId} LIMIT 1`);
    // const post = await db.select(postId);
    res.json(post[0].result);
});

// Route for creating a post
app.post('/post', async (req, res) => {
    const { username, content } = req.body;
    const post = await db.create('post', { username, content });
    res.json(post);
});

// Route for updating a post
app.put('/post/:postId', async (req, res) => {
    const { postId } = req.params;
    const { content } = req.body;
    const post = await db.query('UPDATE $postId SET content = $content', {postId, content});
    res.json(post);
});

// Route for deleting a post
app.delete('/post/:postId', async (req, res) => {
    const { postId } = req.params;
    const post = await db.query(`DELETE $postId`, {postId});
    res.json(post);
});

// Route for commenting on a post
app.post('/comment', async (req, res) => {
    const { postId, username, content } = req.body;
    const comment = await db.create('comment', { postId, username, content });
    res.json(comment);
});

// Route for getting all comments on a post
app.get('/post/:postId/comments', async (req, res) => {
    const { postId } = req.params;
    const comments = await db.query('SELECT * FROM comment WHERE postId = ?', [postId]);
    res.json(comments);
});

// Route for a user to like a post
app.post('/like', async (req, res) => {
    const { username, postId } = req.body;
    // Check if the user and the post exist
    const user = await db.select(username);
    const post = await db.select(postId);
    if (!user || !post) {
        res.status(404).json({ message: 'User or post not found', data: null });
        return;
    }
    // Check if the user has already liked the post
    const existingLike = await db.query(`SELECT * FROM like WHERE in INSIDE $username AND out INSIDE $postId`, {username, postId});
    console.log(existingLike);
    if (existingLike && existingLike.length && existingLike[0].result.length) {
        res.status(400).json({ message: 'User has already liked this post', data: null });
        return;
    }
    // Create the "like" relationship
    await db.query(`RELATE ${username}->like->${postId} UNIQUE SET time.liked = time::now()`);
    res.json({ message: 'Liked post successfully' });
});

// Route for getting all likes on a post
app.get('/post/:postId/likes', async (req, res) => {
    const { postId } = req.params;
    const likes = await db.query('SELECT * FROM user->like->post WHERE post.id = $postId', [postId]);
    res.json(likes);
});

app.listen(4000, () => console.log('Server running on port 4000'));
